*: Ign spec 3 upgrade

yuqi-zhang opened the pull request:
Based on https://github.com/openshift/machine-config-operator/pull/1703, this is a (messy and hacky) MCO that renders all existing configs to spec 3. Currently tested to work with a default install upgrading to this.

Unit and e2e tests not included

yuqi-zhang commented on pkg/apis/machineconfiguration.openshift.io/v1/types.go:
```diff
@@ -96,6 +96,8 @@ type ControllerConfigSpec struct {
 	// regeneration if this changes.
 	NetworkType string `json:"networkType,omitempty"`
+
+	IgnitionSpecVersion int32 `json:"ignitionSpecVersion"`
```
Not needed, will remove

yuqi-zhang commented on pkg/controller/common/helpers.go:
```diff
@@ -250,30 +252,145 @@ func IgnParseWrapper(rawIgn []byte) (interface{}, error) {
+}
+
+// Test: have some situations de-duplicate and see what happens, will merge into main function
+func ParseAndConvertConfigWithDeduplication(rawIgn []byte) (ign3types.Config, error) {
```
will probably merge this into the main function, just trying to think if there's scenarios we won't want to de-duplicate

yuqi-zhang commented on pkg/controller/common/helpers.go:
```diff
@@ -250,30 +252,145 @@ func IgnParseWrapper(rawIgn []byte) (interface{}, error) {
+
+	// TODO jerzhang check if you can have contents and dropins in the same file
+	unitNameMap := map[string]bool{}
+	var outUnits []ign2types.Unit
```
so units is somewhat wonky since it needs to consider dropins when doing a merge, now consider this scenario:
```
Unit A:
  Content: AC
  Dropins:
    Dropin: AD
```
```
Unit B:
    Content: BC
  Dropins:
    Dropin: BD
```
Should the final look like:
```
  Unit rendered:
    Content: BC
  Dropins:
    Dropin: AD
    Dropin: BD
```
or 
```
Unit rendered:
    Content: BC
  Dropins:
    Dropin: BD
```

LorbusChris replied:
I'd be in favor of the former, having both drop ins added.

Another case that may also be considered is if there is a unit in one template, and a dropin only in another template.

yuqi-zhang replied:
In the case of a unit in one and a dropin in another, both the existing spec 2 behaviour and the ignition spec 3 merge function keeps both, so I think we should keep that.

And agreed, having both dropins probably makes it more consistent

yuqi-zhang commented on pkg/controller/render/render_controller.go:
```diff
@@ -481,6 +481,23 @@ func (ctrl *Controller) syncGeneratedMachineConfig(pool *mcfgv1.MachineConfigPoo
+			return err
+		}
+		if needUpdate {
+			_, err = ctrl.client.MachineconfigurationV1().MachineConfigs().Update(context.TODO(), config, metav1.UpdateOptions{})
```
For non-rendered configs I probably need a better location to update them, there is also the Boostrap path to consider

yuqi-zhang replied:
that said with this update in place we will no longer be able to rollback, so maybe for user-created configs I should keep it in the version that it was on?

yuqi-zhang commented on pkg/daemon/daemon.go:
```diff
@@ -1276,17 +1277,33 @@ func (dn *Daemon) validateOnDiskState(currentConfig *mcfgv1.MachineConfig) error
-		return err
+
+	switch typedConfig := ignconfigi.(type) {
+	case ign3types.Config:
```
for validation I'm thinking we should keep validating in the config that it was created in

kikisdeliveryservice replied:
doesnt the ign translator func also validate?

yuqi-zhang replied:
Right in this case I meant validating the on-disk config based on the rendered-config

If the cluster is still on a spec 2 rendered-config version (which shouldn't happen that often, only should happen if a) templates changed causing a race or b) a rendered-config was created during a middle of an update or c) a previous rendered-config update was failing), I'm thinking we should continue to validate in spec 2, instead of translating up and then validating in spec 3.

Maybe we could get away with translating first, just have to be a bit more careful

yuqi-zhang commented:
With https://github.com/openshift/machine-config-operator/pull/1766 rebased this also works on first installs (take a 4.6 installer and override release image with this MCO), which brings up everything in spec 3 automatically.

Next steps:
 - consolidate with new changes in https://github.com/openshift/machine-config-operator/pull/1703
 - clean up code
 - fix tests

LorbusChris commented:
@yuqi-zhang I'll leave https://github.com/openshift/machine-config-operator/pull/1778 open for now, as it might be helpful as a reference. The upgrade path isn't handled there yet, because there is no deduplication of units happening (we'll probably also want this for files).

yuqi-zhang commented:
Updated tests in reference to https://github.com/openshift/machine-config-operator/pull/1778. Probably need to add a few more but hopefully the existing ones pass now.

Also cleaned up the code a bit. Rebased on https://github.com/openshift/machine-config-operator/pull/1703 again and tested manually to pass upgrades + install.

yuqi-zhang commented:
/test e2e-gcp-upgrade

yuqi-zhang commented:
/test e2e-aws

yuqi-zhang commented:
/test e2e-aws

flaking on various other tests, but just to make sure...

yuqi-zhang commented:
Rebased, fixed some tests and fixed unit deduplication a bit more

yuqi-zhang commented:
/retest

runcom commented on pkg/controller/common/helpers.go:
```diff
@@ -271,9 +296,107 @@ func IgnParseWrapper(rawIgn []byte) (ign2types.Config, error) {
+// Append is not considered since we do not allow for appending
+// Units have one exception: dropins are concat'ed
+
+func removeIgnDuplicateFilesAndUnits(ignConfig ign2types.Config) ign2types.Config {
```
I think this function would benefit from a couple of unit tests scenarios 👼 just to make sure I don't go and change something and break some cases - what do you think?

yuqi-zhang replied:
Added a unit test, thanks!

LorbusChris commented on pkg/controller/common/helpers.go:
```diff
@@ -271,9 +296,107 @@ func IgnParseWrapper(rawIgn []byte) (ign2types.Config, error) {
-// internally it transpiles to Ign spec v3 config and translates to spec v2
-func TranspileCoreOSConfigToIgn(files, units []string) (*ign2types.Config, error) {
+// internally it transpiles to Ign spec v3 config
+func TranspileCoreOSConfigToIgn(files, units []string) (*ign3types.Config, error) {
```
Now that we don't convert the FCCT templates to spec v2 anymore, we must ensure that the spec v3 config resulting from this function does not contain duplicate unit and file entries, i.e. they cannot be blindly appended like it has been done until now.

LorbusChris replied:
wondering why we're not hitting this with `kubelet.service` here in CI already (there two templates for it: one that defines a common dropin, and one each for masters/workers that defines the contents)

yuqi-zhang replied:
Edited so we use the updated fcct parser, and parse each file to igntion first and then merge them

yuqi-zhang commented:
/test e2e-aws
/test e2e-gcp-upgrade

yuqi-zhang commented:
/test e2e-vsphere
/test e2e-openstack
/test e2e-ovirt
/test e2e-azure

LorbusChris commented:
/test e2e-azure

runcom commented:
Holding to get https://github.com/openshift/machine-config-operator/pull/1792 first (and I’ll be away tomorrow)

/lgtm
/hold

yuqi-zhang commented:
/test e2e-openstack

LorbusChris commented:
/test e2e-azure
/test e2e-openstack
/test e2e-ovirt
/test e2e-ovn-step-registry
/test e2e-vsphere

yuqi-zhang commented:
Squashed commits and added better message

LorbusChris commented:
/test e2e-vsphere
/test e2e-openstack
/test e2e-ovirt
/test e2e-azure

yuqi-zhang commented:
I don't expect any of those tests to pass except `/test e2e-openstack` , which would be nice to make sure

azure we're out of quota, and the other ones we haven't passed in awhile.

runcom commented:
/lgtm

LorbusChris commented:
:shipit:

kikisdeliveryservice commented:
/test e2e-ovirt

kikisdeliveryservice commented:
/test e2e-gcp-upgrade

kikisdeliveryservice commented:
> I don't expect any of those tests to pass except `/test e2e-openstack` , which would be nice to make sure
> 
> azure we're out of quota, and the other ones we haven't passed in awhile.

Can you run the e2e-azure test on your own azure cluster? or is quota gone for everything?

LorbusChris commented:
now that #1792 has merged, does this need an update to switch the pointer config to v3 as well?
ref: https://github.com/openshift/machine-config-operator/pull/1792/files#diff-5926e0cec606d949b66f823310790298R142

yuqi-zhang commented:
Rebased again and modified stub secret generation to v3

yuqi-zhang commented:
/hold cancel

This is ready to go

> Can you run the e2e-azure test on your own azure cluster? or is quota gone for everything?

Unfortunately the CI account and my account is the same namespace I think

yuqi-zhang commented:
/retest

runcom commented:
/lgtm
